File: cpp/cpp_100-999/768.cc
// C++
// leetcode 768
// https://leetcode.cn/problems/max-chunks-to-make-sorted-ii/

class Solution {
public:

    int maxChunksToSorted(vector<int>& arr) {
    	int n = arr.size();
    	vector<int> mn(n);
    	vector<int> mx(n);
    	mx[0] = arr[0]; mn[n - 1] = arr[n - 1];

    	for (int i = 1; i < n; i++) {
    		mx[i] = max(mx[i - 1], arr[i]);
    		mn[n - 1 - i] = min(mn[n - i], arr[n - 1 - i]);
    	}

    	int res = 1;
    	for (int i = 1; i < n; i++) {
    		if (mx[i - 1] <= mn[i])
    			res++;
    	}
    	return res;
    }
};

File: Go/Go_100-999/768.go
// Go
// leetcode 768
// https://leetcode.cn/problems/max-chunks-to-make-sorted-ii/

func maxChunksToSorted(arr []int) int {
	n := len(arr)
	mx, mn := make([]int, n), make([]int, n)
	mx[0], mn[n-1] = arr[0], arr[n-1]
	for i := 1; i < n; i++ {
		mx[i] = max(mx[i-1], arr[i])
		mn[n-1-i] = min(mn[n-i], arr[n-i-1])
	}

	res := 1
	for i := 1; i < n; i++ {
		if mn[i] >= mx[i-1] {
			res++
		}
	}
	return res
}

func min(a, b int) int {
	if a > b {
		return b
	}
	return a
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}